File: Python3/801 - 1000/821 Shortest Distance to a Character.py
# 44ms
class Solution:
    def shortestToChar(self, S, C):
        """
        :type S: str
        :type C: str
        :rtype: List[int]
        """
        size, last = len(S), -1
        res = []
        for i in range(size):
            if S[i] != C: continue
            if not res:
                res += list(range(i, -1, -1))
            else:
                res += list(range(1, (i - last + 1) // 2)) + list(range((i - last) // 2, -1, -1))
            last = i
        return res + list(range(1, size - last))
    

File: 801 - 1000/821. Shortest Distance to a Character.cc
// 12ms
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> shortestToChar(const string &S, const char C) {
        int size = S.size(), last = -1;
        vector<int> res(size);
        for (int i = 0; i < size; ++i) {
            if (S[i] != C) continue;
            if (last < 0)
                iota(res.rend() - i, res.rend(), 1);
            else {
                int tmp = (i + last) / 2;
                iota(res.begin() + last, res.begin() + tmp, 1);
                iota(res.rend() - i, res.rend() - tmp, 1);
            }
            last = i + 1;
        }
        iota(res.begin() + last, res.end(), 1);
        return res;
    }
};
